### PR TITLE
[runtime] Change the thread wait code to use cond variables instead o…

### PR DESCRIPTION
**Pull request by vargaz**
…f events.

Instead of waiting for the thread handles to be signalled, wait for a cond variable signalled which is signalled by the code which calls pthread_join ().

### CODE REVIEW

File: mono/metadata/threads.c [hunk 1]
```diff
@@ -2966,65 +3025,85 @@ wait_for_tids (struct wait_data *wait, guint32 timeout, gboolean check_state_cha
...
 		mono_threads_unlock ();
 	}
+#endif
+}
+
+static void
+init_wait_data (WaitData *wait)
+{
+	/* This is used calculate the number of joined threads in wait_for_tids () */
+	wait->njoined = njoined_threads;
```
Comment (kumpera) [inline mono/metadata/threads.c]: Won't this be racy because you read njoined_threads without any coordination?

File: mono/metadata/threads.c [hunk 2]
```diff
@@ -2072,6 +2081,16 @@ ves_icall_System_Threading_Thread_MemoryBarrier (void)
...
 }
 
+static void
+signal_background_change (void)
+{
+	thread_wait_lock ();
+	background_changed = TRUE;
+	mono_os_cond_signal (&thread_wait_cond);
+	thread_wait_unlock ();
+	mono_os_event_set (&background_change_event);
```
Comment (kumpera) [inline mono/metadata/threads.c]: Shouldn't background_change_event be gone?

File: mono/metadata/threads.c (hunk 1 above)
Comment (vargaz) [inline mono/metadata/threads.c]: Will fix it.

File: mono/metadata/threads.c (hunk 2 above)
Comment (vargaz) [inline mono/metadata/threads.c]: Will remove it.

Comment (vargaz) [inline mono/metadata/threads.c]: Actually its needs by the windows code.

File: mono/metadata/threads.c [hunk 3]
```diff
@@ -2919,19 +2941,56 @@ static void print_tids (gpointer key, gpointer value, gpointer user)
...
+ * Wait until either:
+ * - wait->num threads are joined
+ * - @check_state_change is TRUE and a thread changes background state
+ * - timeout elapses
+ */
 static void
-wait_for_tids (struct wait_data *wait, guint32 timeout, gboolean check_state_change)
+wait_for_tids (WaitData *wait, guint32 timeout, gboolean check_state_change)
 {
+#ifndef HOST_WIN32
```
Comment (vargaz) [inline mono/metadata/threads.c]: It should be possible to change the windows code to use the same later.

**Comment by kumpera** [comment]
LGTM

**Comment by ludovic-henry** [comment]
I fail to understand how `wait_for_tids` will now wait for a thread to finish. If I understand correctly, wait_for_tids will wake up whenever a thread change it's background state. Also it might be signaled when calling `mono_threads_join_threads`, but that's only called when creating a new thread. But because we do not wait anymore on the `MonoThreadHandle*`, how will it know when one of the thread it's waiting on will be done?

**Comment by ludovic-henry** [comment]
Also is having a different code path between win32 and unix really improving the situation?

**Comment by ludovic-henry** [comment]
And another question, in `wait_for_tids,` how do you know which thread has joined? It is also used by `mono_threads_abort_appdomain_threads` which only tries to abort threads that are in a particular appdomain.

**Comment by vargaz** [comment]
join_threads () is called by the finalizer thread after it is waken up by mono_threads_add_joinable_thread () which is called by thread cleanup code.

As for appdomain_threads, wait_for_tids () only waits for a certain number of threads to exit, but it is called in a loop and each loop iteration recalculates the set of threads to wait.

As for whenever this is a good idea, the wait code needs to wait for threads to really exit, not just their handle being signalled, this was a source of bugs in the past. Also, it removes one use of thread handles, making it possible to remove thread handles completely in the future.

**Comment by ludovic-henry** [comment]
Ok for `mono_threads_join_threads` and `mono_threads_abort_appdomain_threads`, thanks for the explanation.

There has been quite some change on how and when the MonoThreadHandle are signaled. They are now signaled way later than before, guaranteeing that the thread is not going to execute any runtime code anymore. And I am strongly in favor of keeping MonoThreadHandle.

**Pull request merged**